### ISSUE TITLE
Feature request: `filterMap()`

### ISSUE THREAD

**Issue by ndrewtl** [lodash/lodash#5300]
Hi! Thanks so much for maintaining this project. It would be great to have a function to simultaneously map and filter a collection. That is, to map a collection of values to another collection of values, but to exclude a result if the mapping function returns falsy or `null`.

For instance, we could take a set of years, and return the current age of people born before the current year, dropping the years in the future:

```js
const years = [1942, 1978, 2000, 2021, 2022, 2030, 2045];
const currentYear = 2021;
const ages  = filterMap(years, year => year <= currentYear ? currentYear - year : null)
// => [79, 43, 21, 0]
```

A similar feature already exists in [Rust](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map), and I've found it to be quite useful.

**Comment by Scttpr** [comment]
I think you can use the [reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce) method for it in Javascript. For instance :

```js
const years = [1942, 1978, 2000, 2021, 2022, 2030, 2045];
const currentYear = 2021;
const ages = years.reduce((acc, year) => {
  if (year <= currentYear) {
    acc.push(currentYear - year);
  }
  
  return acc;
}, []);
```

**Comment by ndrewtl** [comment]
I agree-- this is how I am currently doing it in my code. However, I would like a separate method. Most methods, even `map()`, could be implemented with `reduce()`, but having a specific method would express the semantics more clearly and simply and clarify my code

**Comment by g-traub** [comment]
You could use [flatMap](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap) :
```js
 const years = [1942, 1978, 2000, 2021, 2022, 2030, 2045];
const currentYear = 2021;
const ages  = years.flatMap(year => year <= currentYear ? [currentYear - year] : [])
// => [79, 43, 21, 0]
```

**Comment by ndrewtl** [comment]
This still wouldn't work in the case that the value I want to return is an array. For instance:

```js
const strings = [
  'Hello world',
  'Lorem ipsum',
  '',
  'Friends, Romans, countrymen',
  'javascript'
]
// Intended value: each string is mapped to a list of words; the empty string is removed
const wordLists = filterMap(strings, str => str.length > 0 ? str.split(' ') : null)
// => [['Hello', 'world'], ['Lorem', 'ipsum'], ['Friends,', 'Romans,', 'countrymen'], ['javascript']]

// Using flatMap: all values are flattened
const flattened= filterMap(strings, str => str.length > 0 ? str.split(' ') :  [])
// => ['Hello', 'world', 'Lorem', 'ipsum', 'Friends,', 'Romans,', 'countrymen', 'javascript']
```

I am not looking to solve the specific example I mentioned above. That example was merely included to demonstrate the intended functionality. As mentioned, I already have a working solution for my need that uses `reduce()`. I am looking for a general implementation of `filterMap()` that can be depended on and used in all cases. The use of this function would also make my code more readable and simpler in the future.

I have implemented `filterMap()` in #5327. I would ask that this PR be merged. If it is merged, I can also contribute Typescript types in the DefinitelyTyped repo before the next release.
